docs/proposals/PROPOSAL_A.md

```diff
@@ -0,0 +1,68 @@
+# Proposal A - Runtime Annotations
+
+This proposal recommends changes that allow runtimes to quickly select the appropriate image from an index using the platform and annotations on the descriptor.
+
+## Links
+
+| Description | Link |
+|-------------|------|
+| Image Index Spec | <https://github.com/opencontainers/image-spec/blob/main/image-index.md> |
+
+## Modifications
+
+### Image Spec
+
+The following language is already included in the Image Index definition:
+
+> If multiple manifests match a client or runtime's requirements, the first matching entry SHOULD be used.
+
+This would add a section for "Annotations":
```

Comment:
In the meetings we talked about a more graph based structure, e.g., where you have top level ownership of some set of annotations. Would that be possible here? I'm worried that, if you have this all flattened, an image supporting multiple kinds of compatibility artifacts could turn into a hugely long list. If you know exactly the name you are looking for (e.g, `archspec.cpu.microarchitecture` then the lookup is O(1), but what happens if you need to get (for some use case) all of the attributes for archspec? You have to go through the list and look for the prefix, likely sorting it first. So I'm wondering if we might do something like:

```
"annotations": {
        "archspec": { 
            "cpu.architecture": "my-favorite-arch",
            ...
    },
        "kernel-thing": { 
            "vegetable": "corn",
             ...
    }
}
```
It's one level more of testing, and allows the user to explicitly index the entire set they are interested in (and get it much more quickly than potentially searching prefixes). Of course this entirely depends on the use case(s) we have in mind.

Reply:
> The annotation key typically follows a reverse domain notation

When talking about compatibility, have to consider parent/child hierarchical relationship, and / or combination relationship

the reverse domain notion is easy to establish hierarchical relationship. it's little hard to figure out how to build the and/or relationship using flat annotations.

for example, the application is compatibility with amd64 architecture and specific nic/gpu requirements:
   1) intel cpu + nic (intel ixgbe nic, mallanox mlx4 or mallanox mlx5) + gpu (nvida telsa v100)   
   2) amd cpu + nic (mallanox mlx6) + gpu (nvida telsa v100 or AMD Instinct MI25)

Reply:
The first idea that comes to my mind.. maybe we could group values under one key to get OR.
For parent/child, as already mentioned by @sudo-bmitch, we could get hierarchy functionality by domain notation.
For example:

Intel CPU:
```
org.opencontainers.image-compatibility.cpu=intel.nic: intel-ixgbe
org.opencontainers.image-compatibility.cpu=intel.nic: mallanox-mlx4
org.opencontainers.image-compatibility.cpu=intel.nic: mallanox-mlx5
org.opencontainers.image-compatibility.cpu=intel.gpu: nvidia-tesla-v100
```

AMD CPU:
```
org.opencontainers.image-compatibility.cpu=amd.nic: mallanox-mlx6
org.opencontainers.image-compatibility.cpu=amd.gpu: nvidia-tesla-v100
org.opencontainers.image-compatibility.cpu=amd.gpu: amd-instinct
```

An alternative would be to have a list of values:
```
org.opencontainers.image-compatibility.cpu=intel.nic: intel-ixgbe mallanox-mlx4 mallanox-mlx5
org.opencontainers.image-compatibility.cpu=intel.gpu: nvidia-tesla-v100
```

Reply:
> This gets pretty hairy looking pretty quickly - it's some kind of selector?

This is not a selector. The idea is that key describe the condition very clearly. Basically it can be any form as long the key on both sides (image index and runtime) match.

> Arguably if two things are functionally the same, this information should be represented somewhere in a graph (e.g., like archspec does) and we shouldn't have to hard code two pieces of the graph into a single annotation. Parsing the graph, if the machine warrants that need to compare, should produce that equality.

As mentioned in the previous comment. I am against transforming the `annotations` field into a nested structure. With this in mind, I came up with examples that are flattened.

Reply:
> The first idea that comes to my mind.. maybe we could group values under one key to get OR.

"Or" is going to be easy since runtimes would ignore annotations they don't recognize. So if multiple entries are listed and the runtime recognizes any of them from it's preference list, that descriptor becomes preferred.

"And" is not so easy, you effectively need to shadow later descriptors with an entry that only has one of the preferred annotations, and then have some kind of ranking or priority of preferences implemented by the runtime to pick a later entry with more than one matching annotation. This adds a level of complication that I wouldn't be surprised to see runtimes decline to implement.

Reply:
From my side this thread can be resolved, for me the proposed solution is good enough.
Waiting for the opinion from @vsoch.

Reply:
okay to summarize my concerns above:

- a single list of annotations, in that it's a flat list, can get potentially large and a tool needs to go through the entire list potentially looking for something.
- there is no ability to understand relationships between things
- you also cannot have multiple attributes per label, perhaps for different contexts

I'm OK to resolve acknowledging the simplicity of the proposal does not address more complex use cases beyond "match this label exactly"

Reply:
This thread is connected to https://github.com/opencontainers/wg-image-compatibility/pull/3#discussion_r1445590268. Before closing I would wait for the opinion from @ChaoyiHuang as he was involved in the discussion too.

> I'm OK to resolve acknowledging the simplicity of the proposal does not address more complex use cases beyond "match this label exactly"

If other proposals would not solve your requirements I would say we can again discuss the approach (but it doesn't mean we have to continue the discussion in this PR).

Reply:
I am fine to merge the proposal, all proposals will do contribute to the final design to some extent.